plc4j/integrations/apache-nifi/nifi-plc4x-processors/src/main/java/org/apache/plc4x/nifi/Plc4xSinkRecordProcessor.java

```diff
@@ -0,0 +1,229 @@
+	}
+
+	@Override
+	public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+		FlowFile fileToProcess = session.get();
+
+        // Abort if there's nothing to do.
+        if (fileToProcess == null) {
+            return;
+        }
+			
+		final ComponentLog logger = getLogger();
+		// Get an instance of a component able to read from a PLC.
+		final AtomicLong nrOfRows = new AtomicLong(0L);
+		final StopWatch executeTime = new StopWatch(true);
+
+		final FlowFile originalFlowFile = fileToProcess;
+
+		InputStream in = session.read(originalFlowFile);
+
+		Record record = null;
+		
+		try (RecordReader recordReader = context.getProperty(PLC_RECORD_READER_FACTORY)
+			.asControllerService(RecordReaderFactory.class)
+			.createRecordReader(originalFlowFile, in, logger)){
+
+			while ((record = recordReader.nextRecord()) != null) {
+				long nrOfRowsHere = 0L;
+				PlcWriteResponse plcWriteResponse;
+				PlcWriteRequest writeRequest;
+
+				Map<String,String> addressMap = getPlcAddressMap(context, fileToProcess);
+				final Map<String, PlcTag> tags = getSchemaCache().retrieveTags(addressMap);
+
+				try (PlcConnection connection = getConnectionManager().getConnection(getConnectionString())) {
+					PlcWriteRequest.Builder builder = connection.writeRequestBuilder();
+					
+					
+					if (tags != null){
+						for (Map.Entry<String,PlcTag> tag : tags.entrySet()){
+							if (record.toMap().containsKey(tag.getKey())) {
+								builder.addTag(tag.getKey(), tag.getValue(), record.getValue(tag.getKey()));
+								nrOfRowsHere++;
+							} else {
+								if (debugEnabled)
+                    				logger.debug("PlcTag " + tag + " is declared as address but was not found on input record.");
+							}
+						}
+					} else {
+						if (debugEnabled)
+                   			logger.debug("Plc-Avro schema and PlcTypes resolution not found in cache and will be added with key: " + addressMap.toString());
+						for (Map.Entry<String,String> entry: addressMap.entrySet()){
+							if (record.toMap().containsKey(entry.getKey())) {
+								builder.addTagAddress(entry.getKey(), entry.getValue(), record.getValue(entry.getKey()));
+								nrOfRowsHere++;
+							} else {
+								if (debugEnabled)
+                    				logger.debug("PlcTag " + entry.getKey() + " with address " + entry.getValue() + " was not found on input record.");
+							}
+						}
+					}
+					writeRequest = builder.build();
+
+					plcWriteResponse = writeRequest.execute().get(
+						context.getProperty(PLC_WRITE_FUTURE_TIMEOUT_MILISECONDS.getName()).asInteger(), TimeUnit.MILLISECONDS
+						);
+				} catch (Exception e) {
+					System.out.println(e.getMessage());
+					in.close();
+					logger.error("Exception writing the data to PLC", e);
+					session.transfer(originalFlowFile, REL_FAILURE);
+					session.commitAsync();
+					throw (e instanceof ProcessException) ? (ProcessException) e : new ProcessException(e);
+				}
+
```

Comment:
should we only cache if it works?

Reply:
I agree, I have moved them in both sinks to cache only if it works